pandas/io/parsers.py

```diff
@@ -458,7 +458,7 @@ def _read(filepath_or_buffer: FilePathOrBuffer, kwds):
     finally:
         parser.close()
 
-    if should_close:
+    if should_close and not isinstance(fp_or_buf, str):
```

Comment:
why is this here and not correctly coming from get_filepath_or_buffer?

Reply:
`get_filepath_or_buffer` can return a str.

as stated before, my preference would be to always add `#type:ignore` rather than change code unnecessarily. The code is currently type safe due to the try/except block, but mypy doesn't see that.

without the isinstance check, `pandas\io\parsers.py:463: error: Item "str" of "Union[str, IO[Any]]" has no attribute "close"`

Reply:
> The code is currently type safe due to the try/except block, but mypy doesn't see that.

sorry i'm wrong here.

it would be type safe if we used `except (AttributeError, ValueError)` but mypy would still report the same error and would need either an ignore or the isinstance check to pass.

at the moment `should_close` is never `True` when `fp_or_buf` is a string. but then again the purpose of type checking is to prevent bugs that may be added in the future.

maybe `get_filepath_or_buffer` should be converted to a context manager and handle the closing.